[Undertow] Add web security and cache-control headers

carterkozak opened the pull request.

Commit:
Add web security and cache-control headers

uschi2000 commented on conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureHandler.java:
```diff
@@ -88,11 +94,35 @@ public ConjureHandler delete(String template, HttpHandler handler) {
+            this.template = template;
+        }
+
+        HttpString getMethod() {
```
IMO, since this is a private class it's fine to omit the getters and just access the fields directly

uschi2000 commented on conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/NoCachingResponseHandler.java:
```diff
@@ -0,0 +1,63 @@
+    private final HttpHandler next;
+
+    NoCachingResponseHandler(HttpHandler next) {
+        this.next = Preconditions.checkNotNull(next, "Delegate HttpHandler is required");
```
IMO fine to avoid this check. We generally assume that everything is non-null unless explicitly documented

uschi2000 commented on conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/NoCachingResponseHandler.java:
```diff
@@ -0,0 +1,63 @@
+        @Override
+        public StreamSinkConduit wrap(ConduitFactory<StreamSinkConduit> factory, HttpServerExchange exchange) {
+            HeaderMap responseHeaders = exchange.getResponseHeaders();
+            if (!responseHeaders.contains(Headers.CACHE_CONTROL)) {
```
should we instead forcefully set the header? under what conditions would it be set already and why would we want to support that configurability?

carterkozak replied:
This could be configured in a manually built endpoint, or if we add a feature similar to the jersey response-as-binary flag. We have tests elsewhere that validate the default can be overridden, but we could ignore those for undertow endpoints if we think that's a bad idea. I don't have a strong preference, but biased toward providing the same functionality.

uschi2000 replied:
I think my preference is for conjure endpoints to be zero config(urability)

uschi2000 commented on conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/WebSecurityHandler.java:
```diff
@@ -0,0 +1,80 @@
+import io.undertow.util.HttpString;
+import org.xnio.conduits.StreamSinkConduit;
+
+final class WebSecurityHandler implements HttpHandler {
```
add high-level docs, maybe a link to the authoritative source of information of how-to-secure-the-web?

uschi2000 commented on conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/WebSecurityHandler.java:
```diff
@@ -0,0 +1,80 @@
+        INSTANCE;
+
+        @Override
+        public StreamSinkConduit wrap(ConduitFactory<StreamSinkConduit> factory, HttpServerExchange exchange) {
```
@walkerburgin could you take a look at this, please?
@cakofony let's also get someone from infosec to take a look?

uschi2000 commented on conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/WebSecurityHandlerTest.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    @Test
+    public void testInternetExploder10() throws IOException {
```
let's avoid throwing punches in public repos

uschi2000 commented:
lgtm modulo some nits, but would appreciate some infosec eyes

Commit:
Merge branch 'develop' into ckozak/additional_handlers

Commit:
Updates based on review

Commit:
Remove check for existing cache-control value

Simplify handlers, there's no need to wait until response streams
are committed.

Commit:
simplify WebSecurityHandler docs

rubenfiszel commented:
We should prefix those PRs with [Undertow] or equivalent

rubenfiszel commented on conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureHandler.java:
```diff
@@ -45,7 +51,7 @@
             // to provide user and trace information on exceptions.
             (endpoint, handler) -> new BearerTokenLoggingHandler(handler),
-            TraceHandler::new,
+            (endpoint, handler) -> new TraceHandler(endpoint.method + " " + endpoint.template, handler),
```
nit: Are there other places where the TraceHandler is instantiated ? Otherwise I would prefer to pass the new EndpointDetails and let the TraceHandler decides how it wants to format it.

carterkozak replied:
EndpointDetails is a private internal structure, the TraceHandler just needs an operation name, I'd prefer not to leak the EndpintDetails abstraction here

rubenfiszel replied:
Could we just pass in the method and template as separated args ? Not strongly opinionated about it but it feel a bit strange in terms of separation of concerns.

carterkozak replied:
I'd like to avoid that, by design I don't want the trace handler to be smart, it should be impossible to make it calculate operation names per request because that can leak unsafe data.

rubenfiszel replied:
👍

rubenfiszel commented on conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureHandler.java:
```diff
@@ -88,11 +94,27 @@ public ConjureHandler delete(String template, HttpHandler handler) {
         return this;
     }
+
+    private static final class EndpointDetails {
```
Seems like the only thing you can do with an EndpointDetails is to display it. That seems like a weird abstraction. Either add getters (or make it immutable) or remove the class in favor of a static method "String displayEndpoint".

carterkozak replied:
We use it for the cache-control handler to avoid handler registration for non-GET requests. @uschi2000 requested that I remove the getters in an earlier review.

rubenfiszel commented on conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/WebSecurityHandlerTest.java:
```diff
@@ -0,0 +1,92 @@
+        server = Undertow.builder()
+                .addHttpListener(12345, "localhost")
+                .setHandler(new WebSecurityHandler(exchange -> {
+                }))
```
inline

Commit:
Merge branch 'develop' into ckozak/additional_handlers

Commit:
WebSecurityHandlerTest formatting

ghost commented:
+1 from AppSec

walkerburgin commented on conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/WebSecurityHandler.java:
```diff
@@ -0,0 +1,80 @@
+            headers.put(Headers.REFERRER_POLICY, REFERRER_POLICY);
+
+            String userAgent = exchange.getRequestHeaders().getFirst(Headers.USER_AGENT);
+            if (userAgent != null) {
```
[This](https://content-security-policy.com/) page suggests that we may run into trouble setting both `content-security-policy` and `x-content-security-policy`: 

> Note: It is known that having both Content-Security-Policy and X-Content-Security-Policy or X-Webkit-CSP causes unexpected behaviours on certain versions of browsers. Please avoid using deprecated X-* headers.

(But if we haven't seen any issues in practice it may not be worth changing here)

uschi2000 replied:
was this resolved?

carterkozak replied:
based on `if we haven't seen any issues in practice it may not be worth changing here` I'd say so. I prefer not to make changes here unless things are currently broken, since this is just applying existing logic in a new format.

walkerburgin commented on conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/WebSecurityHandler.java:
```diff
@@ -0,0 +1,80 @@
+final class WebSecurityHandler implements HttpHandler {
+
+    private static final String CONTENT_SECURITY_POLICY =
+            "default-src 'self'; img-src 'self' data:; style-src 'self' 'unsafe-inline'; frame-ancestors 'self';";
```
`img-src data:` and `style-src 'unsafe-inline'` might be a little too permissive / surprising for defaults. But I guess it may not matter if these headers are being applied predominantly to JSON API endpoints and not page endpoints in practice.

uschi2000 replied:
was this resolved?

carterkozak replied:
I'm happy leaving this as is unless there are serious concerns. If we'd like to lock this down, we should coordinate between all the projects which define this value.

iamdanfox replied:
:+1: this matches the internal content security policy we have in WC already

iamdanfox commented on conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/WebSecurityHandler.java:
```diff
@@ -0,0 +1,82 @@
+                headers.put(HEADER_IE_X_CONTENT_SECURITY_POLICY, CONTENT_SECURITY_POLICY);
+            }
+        }
+        next.handleRequest(exchange);
```
:+1: this block looks exactly the same as WC one we have internally.

Commit:
Merge branch 'develop' into ckozak/additional_handlers
